src/stable_hasher.rs

```diff
@@ -8,43 +8,100 @@ use std::hash::Hasher;
 #[cfg(test)]
 mod tests;
 
-/// Trait for retrieving the result of the stable hashing operation.
-pub trait StableHasherResult: Sized {
-    fn finish(hash: [u64; 2]) -> Self;
-}
-
+/// A Stable Hasher adapted for cross-platform independent hash.
+///
 /// When hashing something that ends up affecting properties like symbol names,
 /// we want these symbol names to be calculated independently of other factors
 /// like what architecture you're compiling *from*.
 ///
 /// To that end we always convert integers to little-endian format before
 /// hashing and the architecture dependent `isize` and `usize` types are
 /// extended to 64 bits if needed.
+///
+/// # Example
+///
+/// ```
+/// use rustc_stable_hash::{StableHasher, StableHasherResult};
+/// use std::hash::Hasher;
+///
+/// struct Hash128([u64; 2]);
+/// impl StableHasherResult for Hash128 {
+///     fn finish(hash: [u64; 2]) -> Hash128 {
+///         Hash128(hash)
+///     }
+/// }
+///
+/// let mut hasher = StableHasher::new();
+/// hasher.write_usize(0xFA);
+///
+/// let hash: Hash128 = hasher.finish();
+/// ```
+#[must_use]
 pub struct StableHasher {
     state: SipHasher128,
 }
 
-impl fmt::Debug for StableHasher {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write!(f, "{:?}", self.state)
-    }
+/// Trait for retrieving the result of the stable hashing operation.
+///
+/// # Example
+///
+/// ```
+/// use rustc_stable_hash::{StableHasher, StableHasherResult};
+///
+/// struct Hash128([u64; 2]);
+///
+/// impl StableHasherResult for Hash128 {
+///     fn finish(hash: [u64; 2]) -> Hash128 {
+///         Hash128(hash)
+///     }
+/// }
+/// ```
```

Comment:
The example could does something more interesting to make it clear why a transformation step might even be needed. E.g. the `Hash128` in the example could contain a `u128`.

Reply:
Indeed, that's a good observation, thks. I've updated the `StableHasherResult` example to be `Hash128(u128)`, as to make it more *interesting*.